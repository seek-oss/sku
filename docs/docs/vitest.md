# Vitest

[Vitest] is a testing framework that supports ESM out-of-the-box, integrates with the Vite ecosystem and has a similar API to Jest.
These features make it a great replacement for Jest in `sku` applications, especially given [Jest's current limitations with ESM].
Sku supports Vitest as an alternate to Jest since v15

**We recommend using vitest with the vite bundler and in an ESM project**, however these are not requirements.

## Migrating to Vitest

Migrating to Vitest should be fairly straightforward.
Before making any changes, it is highly recommend to read [the Vitest documentation](https://vitest.dev/guide/) to familiarise yourself with the API and features, as well as to better understand the differences between Jest and Vitest.

To enable Vitest in `sku`, first install the required dependencies:

```bash
pnpm add -D vitest @sku-lib/vitest
```

Then, configure [`testRunner`][test runner] in your `sku` config:

```typescript
// sku.config.ts
import type { SkuConfig } from 'sku';

export default {
  testRunner: 'vitest',
  ...
} satisfies SkuConfig;
```

`sku` will now invoke [the `vitest` CLI][Vitest CLI] instead of the `jest` CLI when running `sku test`.

#### Key differences between Vitest and Jest

Vitest has strong compatibility with the Jest API, however it still has differences that may affect your tests.
Differences that affect sku projects are listed here.
The full list is documented in the Vitest documentation: [Migrating from Jest to Vitest].

To automate most of the migration process, a codemod is available.
Note that additional changes may still be required after running this codemod.

```sh
pnpm dlx @sku-lib/codemod jest-to-vitest .
```

**Default Watch Mode**

`vitest` defaults to watch mode when running tests.
To run tests without watch mode you can use the `--run` flag:

```json
// package.json
{
  "scripts": {
    // ...
    "test": "sku test --run"
  }
}
```

Watch mode won't trigger in CI environments, so it's safe to omit the flag in your pipeline.

**Code Coverage Dependencies**

Vitest does not install code coverage dependencies by default.
To collect code coverage, pass the `--coverage` flag to the `sku test` command and install the `@vitest/coverage-v8` package:

```bash
pnpm add -D @vitest/coverage-v8
```

**Testing Library Matchers**

If your test setup file includes an import for `@testing-library/jest-dom`, you may need to change this to `@testing-library/jest-dom/vitest`:

```diff
// test-setup.ts
- import '@testing-library/jest-dom';
+ import '@testing-library/jest-dom/vitest';
```

**Globals Disabled**

Jest enables global APIs such as `it`, `describe`, `beforeAll`, etc., by default.
Vitest does not.
Sku adopts the Vitest defaults, meaning you will need to explicitly import these test functions.
This change results in a cleaner global namespace, better type safety, clearer dependencies and is more consistency with modern javascript practices.

```diff
+ import { describe, expect, it } from 'vitest';
```

Be aware that since globals are disabled, some common libraries like `testing-library` will not run auto DOM cleanup.
If using these libraries, you will need to add cleanup to your configured `setupTests` file.

```diff
import '@testing-library/jest-dom/vitest';

+ import { cleanup } from '@testing-library/react';
+ import { afterEach } from 'vitest';

+ afterEach(cleanup)
```

[Vitest]: https://vitest.dev/
[Jest's current limitations with ESM]: https://jestjs.io/docs/ecmascript-modules
[test runner]: ./docs/configuration.md#testrunner
[Vitest CLI]: https://vitest.dev/guide/cli.html
[codemod]: https://codemod.com/registry/jest-vitest
[Migrating from Jest to Vitest]: https://vitest.dev/guide/migration.html#jest
[Vitest globals]: https://vitest.dev/config/#globals
