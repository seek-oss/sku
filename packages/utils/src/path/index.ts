export { toPosixPath } from './toPosixPath.js';
