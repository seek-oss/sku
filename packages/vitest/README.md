[![npm](https://img.shields.io/npm/v/@sku-lib/vitest.svg?style=flat-square)](https://www.npmjs.com/package/@sku-lib/vitest)

# vitest runner for `sku test`
