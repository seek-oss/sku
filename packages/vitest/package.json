{
  "name": "@sku-lib/vitest",
  "version": "0.2.2",
  "description": "vitest runner for sku test",
  "types": "./dist/index.d.ts",
  "main": "./dist/index.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "engines": {
    "node": ">=22.19.0"
  },
  "files": [
    "dist"
  ],
  "type": "module",
  "scripts": {
    "build": "tsc --project tsconfig.build.json",
    "lint:tsc": "tsc --noEmit"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/seek-oss/sku.git",
    "directory": "packages/vite"
  },
  "author": "SEEK",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/seek-oss/sku/issues"
  },
  "homepage": "https://github.com/seek-oss/sku#readme",
  "devDependencies": {
    "typescript-transform-paths": "^3.5.2",
    "vitest": "catalog:"
  },
  "dependencies": {
    "@vanilla-extract/vite-plugin": "catalog:",
    "vite-tsconfig-paths": "^5.1.4"
  },
  "peerDependencies": {
    "vitest": "^3.0.0"
  }
}
