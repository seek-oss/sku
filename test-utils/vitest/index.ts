export { getAppSnapshot } from './appSnapshot.ts';
