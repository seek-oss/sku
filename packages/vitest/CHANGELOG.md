# @sku-lib/vitest

## 0.2.1

### Patch Changes

- Fix `watch` mode ([#1393](https://github.com/seek-oss/sku/pull/1393))

## 0.2.0

### Minor Changes

- Disable Vitest globals ([#1387](https://github.com/seek-oss/sku/pull/1387))

  `vitest` is now a peer dependency. As such, `vitest` APIs must now be imported explicitly

  **EXAMPLE USAGE**:

  ```ts
  // example.test.ts
  import { describe, it, expect } from 'vitest';

  describe("test suite", () => {
    it("should be true" () => {
      expect(true).toBe(true);
    });
  });
  ```

### Patch Changes

- `deps`: Update `vite-plugin-cjs-interop` from `^2.2.0` to `^2.3.0` ([#1386](https://github.com/seek-oss/sku/pull/1386))

- Added missing `compilePackages` and `cjsInteropDependencies` support ([#1381](https://github.com/seek-oss/sku/pull/1381))

- Fixed a bug causing `sku test` to hang when `vitest` is executed in `run` mode ([#1387](https://github.com/seek-oss/sku/pull/1387))

## 0.1.1

### Patch Changes

- Experimental vite bundler now requires vite v7. ([#1346](https://github.com/seek-oss/sku/pull/1346))

  This is a breaking peer dependency change for projects that use the experimental vite bundler. Since this is an experimental feature, we are not bumping the major version.

- - `vitest`: Honour watch mode when running `sku test` ([#1325](https://github.com/seek-oss/sku/pull/1325))

    Previously it was not possible to run Vitest in watch mode via `sku test`. `sku test` should now behave similarly to [the `vitest` CLI][vitest cli], so `sku test`, `sku test watch` and `sku test -w/--watch` will all trigger watch mode.

    To run tests once and exit, you can execute `sku test run`, `sku test --run` or `sku test --no-watch` .

    [vitest cli]: https://vitest.dev/guide/cli

  - `deps` Bump peer dependencies to support React 19

## 0.1.0

### Minor Changes

- Initial release. ([#1273](https://github.com/seek-oss/sku/pull/1273))
