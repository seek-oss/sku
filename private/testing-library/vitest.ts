// Re-export vitest setup from cli-testing-library
import 'cli-testing-library/vitest';
