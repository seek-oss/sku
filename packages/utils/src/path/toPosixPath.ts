import path from 'node:path';

/**
 * Replaces all win32 path separators (\) with posix path separators (/)
 */
export const toPosixPath = (inputPath: string) =>
  inputPath.replaceAll(path.win32.sep, path.posix.sep);
