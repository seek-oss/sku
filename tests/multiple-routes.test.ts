import { describe, beforeAll, afterAll, it } from 'vitest';
import path from 'node:path';
import {
  dirContentsToObject,
  getPort,
  runSkuScriptInDir,
  waitForUrls,
  createCancelSignal,
} from '@sku-private/test-utils';
import { getAppSnapshot } from '@sku-private/vitest-utils';

import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);

const appDir = path.dirname(
  require.resolve('@sku-fixtures/multiple-routes/sku.config.js'),
);

const targetDirectory = `${appDir}/dist`;

const renderPageCorrectly = async ({
  page,
  pageUrl,
}: {
  page: string;
  pageUrl: string;
}) => {
  it(`should render ${page} page correctly`, async ({ expect }) => {
    const snapshot = await getAppSnapshot({
      url: pageUrl,
      expect,
      waitUntil: 'networkidle0',
    });
    expect(snapshot).toMatchSnapshot();
  });
};

describe('multiple-routes', () => {
  describe.sequential.for(['vite'])('bundler: %s', (bundler) => {
    describe('start', async () => {
      const { cancel, signal } = createCancelSignal();

      const port = await getPort();

      const url = `http://localhost:${port}`;
      const args = ['--strict-port', `--port=${port}`];

      if (bundler === 'vite') {
        args.push(
          '--convert-loadable',
          '--experimental-bundler',
          '--config',
          'sku.config.vite.js',
        );
      }

      beforeAll(async () => {
        runSkuScriptInDir('start', appDir, {
          args,
          signal,
        });
        await waitForUrls(url);
      });

      afterAll(async () => {
        cancel();
      });

      renderPageCorrectly({
        page: 'home',
        pageUrl: url,
      });

      renderPageCorrectly({
        page: 'details',
        pageUrl: `${url}/details/123`,
      });
    });

    describe('build and serve', async () => {
      const { cancel, signal } = createCancelSignal();

      const port = await getPort();

      const url = `http://localhost:${port}`;
      const portArgs = ['--strict-port', `--port=${port}`];

      const args: string[] = [];

      if (bundler === 'vite') {
        args.push(
          '--convert-loadable',
          '--experimental-bundler',
          '--config',
          'sku.config.vite.js',
        );
      }

      beforeAll(async () => {
        await runSkuScriptInDir('build', appDir, { args });
        runSkuScriptInDir('serve', appDir, {
          args: portArgs,
          signal,
        });
        await waitForUrls(url);
      });

      afterAll(async () => {
        cancel();
      });

      renderPageCorrectly({
        page: 'home',
        pageUrl: url,
      });

      renderPageCorrectly({
        page: 'details',
        pageUrl: `${url}/details/123`,
      });

      it('should generate the expected files', async ({ expect }) => {
        const files = await dirContentsToObject(targetDirectory);
        expect(files).toMatchSnapshot();
      });
    });
  });

  describe('test', () => {
    it('should handle dynamic imports in tests', async ({ expect }) => {
      await expect(
        runSkuScriptInDir('test', appDir),
      ).resolves.not.toThrowError();
    });
  });
});
