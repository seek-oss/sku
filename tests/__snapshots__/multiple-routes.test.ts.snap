// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`multiple-routes > bundler: vite > build and serve > should generate the expected files 1`] = `
{
  "AsyncComponent-BbylO-Ju.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "Details-CNUAR5xk.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "Details-DGJnH813.css": "._1hpn1da0{color:#00f}
",
  "Home-D-106MY_.css": ".ms22la0{color:red}
",
  "Home-DQBwRi9y.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "production/au/details/$id/index.html": "<!DOCTYPE html>
<html>
  <head><meta http-equiv="Content-Security-Policy" content="script-src 'self' https://error-tracking.com https://fb-tracking.com https://code.jquery.com 'sha256-rGv16tw6EM5EMe4gsxdgzpqzrzd14bhfa2neQclOXDU=';">
    <meta charset="UTF-8" >
    <title>hello-world</title>
    <meta name="viewport" content="width=device-width, initial-scale=1" >
    <script src="https://code.jquery.com/jquery-3.5.0.slim.min.js"></script>
    <script>
      console.log('Hi');
    </script>
    <style type="text/css">
      body {
        background: pink;
      }
    </style>
    <link rel="stylesheet" href="/static/place/Details-DGJnH813.css" crossorigin >

<link rel="modulepreload" href="/static/place/vite-client.js" crossorigin >

<link rel="modulepreload" href="/static/place/Details-CNUAR5xk.js" crossorigin >

<link rel="modulepreload" href="/static/place/AsyncComponent-BbylO-Ju.js" crossorigin >

  </head>
  <body>
    <div id="app"><!--$--><h1 class="_1hpn1da0">Welcome to the Details page - au<span>Some special async content</span></h1><!--/$--></div>
    <script id="__SKU_CLIENT_CONTEXT__" type="application/json">{"site":"au"}</script>
<script type="module" src="/static/place/AsyncComponent-BbylO-Ju.js" ></script>
<script type="module" src="/static/place/Details-CNUAR5xk.js" ></script>
<script type="module" src="/static/place/vite-client.js" ></script>
  </body>
</html>",
  "production/au/index.html": "<!DOCTYPE html>
<html>
  <head><meta http-equiv="Content-Security-Policy" content="script-src 'self' https://error-tracking.com https://fb-tracking.com https://code.jquery.com 'sha256-rGv16tw6EM5EMe4gsxdgzpqzrzd14bhfa2neQclOXDU=';">
    <meta charset="UTF-8" >
    <title>hello-world</title>
    <meta name="viewport" content="width=device-width, initial-scale=1" >
    <script src="https://code.jquery.com/jquery-3.5.0.slim.min.js"></script>
    <script>
      console.log('Hi');
    </script>
    <style type="text/css">
      body {
        background: pink;
      }
    </style>
    <link rel="stylesheet" href="/static/place/Home-D-106MY_.css" crossorigin >

<link rel="modulepreload" href="/static/place/vite-client.js" crossorigin >

<link rel="modulepreload" href="/static/place/Home-DQBwRi9y.js" crossorigin >

  </head>
  <body>
    <div id="app"><!--$--><h1 class="ms22la0">Welcome to the Home page - au</h1><!--/$--></div>
    <script id="__SKU_CLIENT_CONTEXT__" type="application/json">{"site":"au"}</script>
<script type="module" src="/static/place/Home-DQBwRi9y.js" ></script>
<script type="module" src="/static/place/vite-client.js" ></script>
  </body>
</html>",
  "production/nz/nz/details/:id/index.html": "<!DOCTYPE html>
<html>
  <head><meta http-equiv="Content-Security-Policy" content="script-src 'self' https://error-tracking.com https://fb-tracking.com https://code.jquery.com 'sha256-rGv16tw6EM5EMe4gsxdgzpqzrzd14bhfa2neQclOXDU=';">
    <meta charset="UTF-8" >
    <title>hello-world</title>
    <meta name="viewport" content="width=device-width, initial-scale=1" >
    <script src="https://code.jquery.com/jquery-3.5.0.slim.min.js"></script>
    <script>
      console.log('Hi');
    </script>
    <style type="text/css">
      body {
        background: pink;
      }
    </style>
    <link rel="stylesheet" href="/static/place/Details-DGJnH813.css" crossorigin >

<link rel="modulepreload" href="/static/place/vite-client.js" crossorigin >

<link rel="modulepreload" href="/static/place/Details-CNUAR5xk.js" crossorigin >

<link rel="modulepreload" href="/static/place/AsyncComponent-BbylO-Ju.js" crossorigin >

  </head>
  <body>
    <div id="app"><!--$--><h1 class="_1hpn1da0">Welcome to the Details page - nz<span>Some special async content</span></h1><!--/$--></div>
    <script id="__SKU_CLIENT_CONTEXT__" type="application/json">{"site":"nz"}</script>
<script type="module" src="/static/place/AsyncComponent-BbylO-Ju.js" ></script>
<script type="module" src="/static/place/Details-CNUAR5xk.js" ></script>
<script type="module" src="/static/place/vite-client.js" ></script>
  </body>
</html>",
  "production/nz/nz/index.html": "<!DOCTYPE html>
<html>
  <head><meta http-equiv="Content-Security-Policy" content="script-src 'self' https://error-tracking.com https://fb-tracking.com https://code.jquery.com 'sha256-rGv16tw6EM5EMe4gsxdgzpqzrzd14bhfa2neQclOXDU=';">
    <meta charset="UTF-8" >
    <title>hello-world</title>
    <meta name="viewport" content="width=device-width, initial-scale=1" >
    <script src="https://code.jquery.com/jquery-3.5.0.slim.min.js"></script>
    <script>
      console.log('Hi');
    </script>
    <style type="text/css">
      body {
        background: pink;
      }
    </style>
    <link rel="stylesheet" href="/static/place/Home-D-106MY_.css" crossorigin >

<link rel="modulepreload" href="/static/place/vite-client.js" crossorigin >

<link rel="modulepreload" href="/static/place/Home-DQBwRi9y.js" crossorigin >

  </head>
  <body>
    <div id="app"><!--$--><h1 class="ms22la0">Welcome to the Home page - nz</h1><!--/$--></div>
    <script id="__SKU_CLIENT_CONTEXT__" type="application/json">{"site":"nz"}</script>
<script type="module" src="/static/place/Home-DQBwRi9y.js" ></script>
<script type="module" src="/static/place/vite-client.js" ></script>
  </body>
</html>",
  "vite-client.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`multiple-routes > bundler: vite > build and serve > should render details page correctly 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta
      http-equiv="Content-Security-Policy"
      content="script-src 'self' https://error-tracking.com https://fb-tracking.com https://code.jquery.com 'sha256-rGv16tw6EM5EMe4gsxdgzpqzrzd14bhfa2neQclOXDU=';"
    >
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <script src="https://code.jquery.com/jquery-3.5.0.slim.min.js">
    </script>
    <script>
      console.log('Hi');
    </script>
    <style type="text/css">
      body {
        background: pink;
      }
    </style>
    <link
      rel="stylesheet"
      href="/static/place/Details-DGJnH813.css"
      crossorigin
    >
    <link
      rel="modulepreload"
      href="/static/place/vite-client.js"
      crossorigin
    >
    <link
      rel="modulepreload"
      href="/static/place/Details-CNUAR5xk.js"
      crossorigin
    >
    <link
      rel="modulepreload"
      href="/static/place/AsyncComponent-BbylO-Ju.js"
      crossorigin
    >
  </head>
  <body>
    <div id="app">
      <h1 class="_1hpn1da0">
        Welcome to the Details page - au
        <span>
          Some special async content
        </span>
      </h1>
    </div>
    <script
      id="__SKU_CLIENT_CONTEXT__"
      type="application/json"
    >
      {"site":"au"}
    </script>
    <script
      type="module"
      src="/static/place/AsyncComponent-BbylO-Ju.js"
    >
    </script>
    <script
      type="module"
      src="/static/place/Details-CNUAR5xk.js"
    >
    </script>
    <script
      type="module"
      src="/static/place/vite-client.js"
    >
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: 
===================================================================
--- sourceHtml
+++ clientHtml
@@ -47,7 +47,7 @@
   <body>
     <div id="app">
       <h1 class="_1hpn1da0">
-        Welcome to the Details page - au
+        Welcome to the Details page - auID: 123
         <span>
           Some special async content
         </span>
`;

exports[`multiple-routes > bundler: vite > build and serve > should render home page correctly 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta
      http-equiv="Content-Security-Policy"
      content="script-src 'self' https://error-tracking.com https://fb-tracking.com https://code.jquery.com 'sha256-rGv16tw6EM5EMe4gsxdgzpqzrzd14bhfa2neQclOXDU=';"
    >
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <script src="https://code.jquery.com/jquery-3.5.0.slim.min.js">
    </script>
    <script>
      console.log('Hi');
    </script>
    <style type="text/css">
      body {
        background: pink;
      }
    </style>
    <link
      rel="stylesheet"
      href="/static/place/Home-D-106MY_.css"
      crossorigin
    >
    <link
      rel="modulepreload"
      href="/static/place/vite-client.js"
      crossorigin
    >
    <link
      rel="modulepreload"
      href="/static/place/Home-DQBwRi9y.js"
      crossorigin
    >
  </head>
  <body>
    <div id="app">
      <h1 class="ms22la0">
        Welcome to the Home page - au
      </h1>
    </div>
    <script
      id="__SKU_CLIENT_CONTEXT__"
      type="application/json"
    >
      {"site":"au"}
    </script>
    <script
      type="module"
      src="/static/place/Home-DQBwRi9y.js"
    >
    </script>
    <script
      type="module"
      src="/static/place/vite-client.js"
    >
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: NO DIFF
`;

exports[`multiple-routes > bundler: vite > start > should render details page correctly 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta
      http-equiv="Content-Security-Policy"
      content="script-src 'self' https://error-tracking.com https://fb-tracking.com https://code.jquery.com 'sha256-NEZvGkT0ZWP6XHdKYM4B1laRPcM6Lw4LJfkDtIEVAKc=' 'sha256-rGv16tw6EM5EMe4gsxdgzpqzrzd14bhfa2neQclOXDU=';"
    >
    <script type="module">
      import { injectIntoGlobalHook } from "/@react-refresh"
injectIntoGlobalHook(window);
window.$RefreshReg$ = () => {};
window.$RefreshSig$ = () => (type) => type;
    </script>
    <script
      type="module"
      src="/@vite/client"
    >
    </script>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <script src="https://code.jquery.com/jquery-3.5.0.slim.min.js">
    </script>
    <script>
      console.log('Hi');
    </script>
    <style type="text/css">
      body {
        background: pink;
      }
    </style>
    <script
      type="module"
      src="/@id/__x00__/index.html?html-proxy&index=0.js"
    >
    </script>
    <script
      type="module"
      src="/@id/__x00__/index.html?html-proxy&index=1.js"
    >
    </script>
  </head>
  <body>
    <div id="app">
      <h1 class="Details_root__1hpn1da0">
        Welcome to the Details page - au
        <span>
          Some special async content
        </span>
      </h1>
    </div>
    <script
      type="module"
      src="{cwd}/packages/sku/dist/services/vite/entries/vite-client.js"
    >
    </script>
    <script
      id="__SKU_CLIENT_CONTEXT__"
      type="application/json"
    >
      {"site":"au"}
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: 
===================================================================
--- sourceHtml
+++ clientHtml
@@ -36,19 +36,27 @@
     </style>
     <script
       type="module"
-      src="/@id/__x00__/index.html?html-proxy&index=0.js"
+      src="/@id/__x00__/index.html?html-proxy&amp;index=0.js"
     >
     </script>
     <script
       type="module"
-      src="/@id/__x00__/index.html?html-proxy&index=1.js"
+      src="/@id/__x00__/index.html?html-proxy&amp;index=1.js"
     >
     </script>
+    <style
+      type="text/css"
+      data-vite-dev-id="{cwd}/fixtures/multiple-routes/src/handlers/Details.css.js.vanilla.css"
+    >
+      .Details_root__1hpn1da0 {
+  color: blue;
+}
+    </style>
   </head>
   <body>
     <div id="app">
       <h1 class="Details_root__1hpn1da0">
-        Welcome to the Details page - au
+        Welcome to the Details page - auID: 123
         <span>
           Some special async content
         </span>
`;

exports[`multiple-routes > bundler: vite > start > should render home page correctly 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta
      http-equiv="Content-Security-Policy"
      content="script-src 'self' https://error-tracking.com https://fb-tracking.com https://code.jquery.com 'sha256-NEZvGkT0ZWP6XHdKYM4B1laRPcM6Lw4LJfkDtIEVAKc=' 'sha256-rGv16tw6EM5EMe4gsxdgzpqzrzd14bhfa2neQclOXDU=';"
    >
    <script type="module">
      import { injectIntoGlobalHook } from "/@react-refresh"
injectIntoGlobalHook(window);
window.$RefreshReg$ = () => {};
window.$RefreshSig$ = () => (type) => type;
    </script>
    <script
      type="module"
      src="/@vite/client"
    >
    </script>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <script src="https://code.jquery.com/jquery-3.5.0.slim.min.js">
    </script>
    <script>
      console.log('Hi');
    </script>
    <style type="text/css">
      body {
        background: pink;
      }
    </style>
    <script
      type="module"
      src="/@id/__x00__/index.html?html-proxy&index=0.js"
    >
    </script>
    <script
      type="module"
      src="/@id/__x00__/index.html?html-proxy&index=1.js"
    >
    </script>
  </head>
  <body>
    <div id="app">
      <h1 class="Home_root__ms22la0">
        Welcome to the Home page - au
      </h1>
    </div>
    <script
      type="module"
      src="{cwd}/packages/sku/dist/services/vite/entries/vite-client.js"
    >
    </script>
    <script
      id="__SKU_CLIENT_CONTEXT__"
      type="application/json"
    >
      {"site":"au"}
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: 
===================================================================
--- sourceHtml
+++ clientHtml
@@ -36,14 +36,22 @@
     </style>
     <script
       type="module"
-      src="/@id/__x00__/index.html?html-proxy&index=0.js"
+      src="/@id/__x00__/index.html?html-proxy&amp;index=0.js"
     >
     </script>
     <script
       type="module"
-      src="/@id/__x00__/index.html?html-proxy&index=1.js"
+      src="/@id/__x00__/index.html?html-proxy&amp;index=1.js"
     >
     </script>
+    <style
+      type="text/css"
+      data-vite-dev-id="{cwd}/fixtures/multiple-routes/src/handlers/Home.css.js.vanilla.css"
+    >
+      .Home_root__ms22la0 {
+  color: red;
+}
+    </style>
   </head>
   <body>
     <div id="app">
`;
